service/forward_service.cc

```diff
@@ -201,20 +201,22 @@ future<query::forward_result> forward_service::dispatch_to_shards(
 ) {
     _stats.requests_dispatched_to_own_shards += 1;
 
-    auto result = co_await container().map_reduce0(
+    return container().map_reduce0(
         [req, tr_info] (auto& fs) {
             return fs.execute_on_this_shard(req, tr_info);
         },
         std::optional<query::forward_result>(),
-        [&req] (std::optional<query::forward_result> partial, query::forward_result mapped) -> std::optional<query::forward_result>{
+        [reduction_types = req.reduction_types] (std::optional<query::forward_result> partial, query::forward_result mapped) -> std::optional<query::forward_result>{
             if (partial) {
-                mapped.merge(*partial, req.reduction_types);
+                mapped.merge(*partial, reduction_types);
             }
             return {mapped};
         }
+    ).then(
+        [] (std::optional<query::forward_result> result) {
+            return *result;
```

Comment:
Would it be hard to instead return an always-engaged optional from this function and dereference it in all call sites? That way we save 1 allocation for this `then()` chain.

Reply:
I guess once per forward request is ok. Hopefully one they we can roll the coroutine back, and meanwhile it can stay as is

Reply:
Isn't merging `mapped` into a newly constructed `query::forward_result` equivalent to copying `mapped`?
If so, I don't think we don't need the whole std::optional spiel.

Reply:
or sorry, it's the other way around, but similar question.  why not call `mapped.merge(partial, reduction_types)` unconditionally (also when it's empty after begin default constructed)

Reply:
At first i thought that `forward_result::merge` has to work with non-empty results, but this is not the case.

https://github.com/scylladb/scylla/blob/5d59c841d0afaae888c9c5ed7db6aeb54969f4f9/query.cc#L405-L407

It's a bit unintuitive - `mapped.merge(partial, reduction_types) ` would not work, but `partial.merge(mapped, reduction_types)` is ok. Current implementation allow merging empty result only when `this` is empty. Other case of merging (`this` is not empty, `other` is empty) results in an error. It would probably be good to change the behavior of `forward_result::merge` to allow merging in that other case too.

Thanks for your suggestion, I will remove `std::optional` it in the next version.